README
This is optipng-ant-task, an Apache Ant task for optimizing PNG images.
Under the hood this task uses (and requires) OptiPNG, see:

   http://optipng.sourceforge.net/

This Ant task has been tested with the following combination of software:

   - J2SE 5.0 and Java SE 6 (source code uses Java 5-features)
   - Ant 1.7.1
   - OptiPNG 0.6.3

This software is available under the terms of a BSD-style license, see
the accompanied LICENSE file.

This task comes as a collection of Java source files that can be built with
Ant. To compile the Java classes, run all available unit tests and build the
JAR file run:

   ant

Note that this already requires 'optipng' on the PATH. The result will be a
JAR file:

   build/optipng-ant-task.jar

If you don't have 'optipng' installed and still want to build the JAR file,
just execute:

   ant jar

This will skip the execution of the available unit tests.

Example usage of the task in an Ant build file:

   <taskdef name="optipng"
       classname="com.pensioenpage.jynx.optipng.OptiPNGTask"
       classpath="lib/optipng-ant-task.jar" />

   <optipng dir="src/htdocs" todir="build/htdocs" />

Although all parameters are optional, the task supports various:

   process   - flag that indicates if the input files should be processed at
               all; there are 3 possible modes:

                  "yes" - the files must be processed by OptiPNG, if at least
                          one file could not be processed, then this task will
                          fail (alternative is "true");

                  "no"  - the files must not be processed by OptiPNG but must
                          instead just be copied as-is, unchanged (alternative
                          is "false");

                  "try" - attempt to process all matching file(s), when a file
                          fails to be processed, then copy the original file
                          instead.

   dir       - the source directory, to read PNG image files from, defaults to
               the project base directory;

   todir     - the destination directory, to write the generated CSS files to,
               defaults to the source directory;

   overwrite - when set this flag forces each existing output file to be
               overwritten, even if it is newer than the respective source
               file; defaults to 'no';

   command   - the command to execute, by default the task uses 'optipng';

   timeOut   - the time-out in milliseconds for executing a single command,
               defaults to 60000 (meaning 60 seconds);

   includes  - the files in the source directory to include, defaults to all
               files, although only those that end in '.png' will actually
               be optimized or copied;

   excludes  - the files to exclude, even if they are matched by the includes;

and other parameters inherited from the MatchingTask, see:

   http://ant.apache.org/manual/dirtasks.html

If you want to file a bug report or a feature request, please do so here:

   http://github.com/znerd/optipng-ant-task/issues

This software has been developed by:

   Ernst de Haan      -  ernst@pensioenpage.com
   PensioenPage B.V.  -  https://www.pensioenpage.com/


RELEASES
---- VERSION 0.2 (September 7, 2009) -----------------------------------------

Implemented 'process' option.
Implemented 'overwrite' option.
Fixed broken support for 'includes' parameter.
Improved logging.
Added 2nd unit test.


---- VERSION 0.1 (September 7, 2009) -----------------------------------------

Initial release.


src/com/pensioenpage/jynx/optipng/Buffer.java
// Copyright 2007-2009, PensioenPage B.V.
package com.pensioenpage.jynx.optipng;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.IOException;
import java.io.OutputStream;

import org.apache.tools.ant.taskdefs.ExecuteStreamHandler;
import org.apache.tools.ant.taskdefs.StreamPumper;

/**
 * An <code>ExecuteStreamHandler</code> implementation that stores all
 * output in a buffer.
 *
 * @version $Revision: 10190 $ $Date: 2009-08-25 17:49:35 +0200 (di, 25 aug 2009) $
 * @author <a href="mailto:ernst@pensioenpage.com">Ernst de Haan</a>
 */
class Buffer extends Object implements ExecuteStreamHandler {

   // TODO: Add state checking

   //-------------------------------------------------------------------------
   // Constructors
   //-------------------------------------------------------------------------

   /**
    * Constructs a new <code>Buffer</code>.
    */
   public Buffer() {
      _outBuffer = new ByteArrayOutputStream();
      _errBuffer = new ByteArrayOutputStream();
   }


   //-------------------------------------------------------------------------
   // Fields
   //-------------------------------------------------------------------------

   /**
    * The buffer holding all <em>stdout</em> output. Never <code>null</code>.
    */
   private final ByteArrayOutputStream _outBuffer;

   /**
    * The buffer holding all <em>stderr</em> output. Never <code>null</code>.
    */
   private final ByteArrayOutputStream _errBuffer;

   /**
    * The thread pumping the <em>stdout</em> output to the buffer.
    * Initially <code>null</code>, set by
    * {@link #setProcessOutputStream(InputStream)}.
    * The thread is started from the {@link #start()} method.
    */
   private Thread _outThread;

   /**
    * The thread pumping the <em>stderr</em> output to the buffer.
    * Initially <code>null</code>, set by
    * {@link #setProcessErrorStream(InputStream)}.
    * The thread is started from the {@link #start()} method.
    */
   private Thread _errThread;


   //----------------------------------------------------------------------
   // Methods
   //----------------------------------------------------------------------

   // Specified by ExecuteStreamHandler
   public void setProcessInputStream(OutputStream os) {
      // ignore, we don't send input to the process
   }

   // Specified by ExecuteStreamHandler
   public void setProcessOutputStream(InputStream is) {
      _outThread = new Thread(new StreamPumper(is, _outBuffer));
   }

   // Specified by ExecuteStreamHandler
   public void setProcessErrorStream(InputStream is) {
      _errThread = new Thread(new StreamPumper(is, _errBuffer));
   }

   // Specified by ExecuteStreamHandler
   public void start() {
      _outThread.start();
      _errThread.start();
   }

   // Specified by ExecuteStreamHandler
   public void stop() {
      try {
         _outThread.join();
      } catch (InterruptedException e) {
         // ignore
      }
      try {
         _errThread.join();
      } catch (InterruptedException e) {
         // ignore
      }
   }

   /**
    * Copied all collected <em>stdout</em> output to the specified output
    * stream.
    *
    * @param os
    *    the {@link OutputStream} to copy the collected <em>stdout</em> output
    *    to, cannot be <code>null</code>.
    *
    * @throws IllegalArgumentException
    *    if <code>os == null</code>.
    *
    * @throws IOException
    *    in case of an I/O error.
    */
   public void writeOutTo(OutputStream os)
   throws IOException {
      if (os == null) {
         throw new IllegalArgumentException("os == null");
      }
      _outBuffer.writeTo(os);
   }

   /**
    * Retrieves all collected <em>stdout</em> output as a character string.
    * The platform default encoding is used to convert the underlying bytes to
    * characters.
    *
    * @return
    *    the collected <em>stdout</em> output as a character string,
    *    never <code>null</code>.
    */
   public String getOutString() {
      return _outBuffer.toString();
   }

   /**
    * Copied all collected <em>stderr</em> output to the specified output
    * stream.
    *
    * @param os
    *    the {@link OutputStream} to copy the collected <em>stderr</em> output
    *    to, cannot be <code>null</code>.
    *
    * @throws IllegalArgumentException
    *    if <code>os == null</code>.
    *
    * @throws IOException
    *    in case of an I/O error.
    */
   public void writeErrTo(OutputStream os) throws IOException {
      _errBuffer.writeTo(os);
   }

   /**
    * Retrieves all collected <em>stderr</em> output as a character string.
    * The platform default encoding is used to convert the underlying bytes to
    * characters.
    *
    * @return
    *    the collected <em>stderr</em> output as a character string,
    *    never <code>null</code>.
    */
   public String getErrString() {
      return _errBuffer.toString();
   }
}


src/com/pensioenpage/jynx/optipng/OptiPNGTask.java
// Copyright 2007-2009, PensioenPage B.V.
package com.pensioenpage.jynx.optipng;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.IOException;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;

import org.apache.tools.ant.BuildException;
import org.apache.tools.ant.DirectoryScanner;
import org.apache.tools.ant.Project;
import static org.apache.tools.ant.Project.MSG_ERR;
import static org.apache.tools.ant.Project.MSG_VERBOSE;
import org.apache.tools.ant.taskdefs.Execute;
import org.apache.tools.ant.taskdefs.ExecuteStreamHandler;
import org.apache.tools.ant.taskdefs.ExecuteWatchdog;
import org.apache.tools.ant.taskdefs.MatchingTask;
import org.apache.tools.ant.types.FileSet;
import org.apache.tools.ant.util.FileUtils;

/**
 * An Apache Ant task for optimizing a number of PNG files, using OptiPNG. For
 * more information, see the
 * <a href="http://optipng.sourceforge.net/">OptiPNG homepage</a>.
 *
 * <p>The most notable parameters supported by this task are:
 *
 * <dl>
 * <dt>command
 * <dd>The name of the command to execute.
 *     Optional, defaults to <code>optipng</code>.
 *
 * <dt>timeOut
 * <dd>The time-out for each individual invocation of the command, in
 *     milliseconds. Optional, defaults to 60000 (60 seconds).
 *
 * <dt>dir
 * <dd>The source directory to read from.
 *     Optional, defaults to the project base directory.
 *
 * <dt>includes
 * <dd>The files to match in the source directory.
 *     Optional, defaults to all files.
 *
 * <dt>overwrite
 * <dd>Flag that can be used to force overwriting of existing newer files.
 *     Optional, defaults to <em>false</em>.
 *
 * <dt>excludes
 * <dd>The files to exclude, even if they are matched by the include filter.
 *     Optional, default is empty.
 *
 * <dt>toDir
 * <dd>The target directory to write to.
 *     Optional, defaults to the source directory.
 * </dl>
 *
 * <p>This task supports more parameters and contained elements, inherited
 * from {@link MatchingTask}. For more information, see
 * <a href="http://ant.apache.org/manual/dirtasks.html">the Ant site</a>.
 *
 * @author <a href="mailto:ernst@pensioenpage.com">Ernst de Haan</a>
 */
public final class OptiPNGTask extends MatchingTask {

   //-------------------------------------------------------------------------
   // Class fields
   //-------------------------------------------------------------------------

   /**
    * The name of the default OptiPNG command: <code>"optipng"</code>.
    */
   public static final String DEFAULT_COMMAND = "optipng";

   /**
    * The default time-out: 60 seconds.
    */
   public static final long DEFAULT_TIMEOUT = 60L * 1000L;


   //-------------------------------------------------------------------------
   // Class functions
   //-------------------------------------------------------------------------

   /**
    * Returns a quoted version of the specified string,
    * or <code>"(null)"</code> if the argument is <code>null</code>.
    *
    * @param s
    *    the character string, can be <code>null</code>,
    *    e.g. <code>"foo bar"</code>.
    *
    * @return
    *    the quoted string, e.g. <code>"\"foo bar\""</code>,
    *    or <code>"(null)"</code> if the argument is <code>null</code>.
    */
   private static final String quote(String s) {
      return s == null ? "(null)" : "\"" + s + '"';
   }

   /**
    * Returns a quoted version string representation,
    * or <code>"(null)"</code> if the argument is <code>null</code>.
    *
    * @param o
    *    the object, can be <code>null</code>.
    *
    * @return
    *    the quoted string representation of the specified object,
    *    e.g. <code>"\"foo bar\""</code>,
    *    or <code>"(null)"</code> if the argument is <code>null</code>.
    */
   private static final String quote(Object o) {
      return o == null ? "(null)" : quote(o.toString());
   }

   /**
    * Determines if the specified character string matches the regular
    * expression.
    *
    * @param s
    *    the string to research, or <code>null</code>.
    *
    * @param regex
    *    the regular expression, cannot be <code>null</code>.
    *
    * @return
    *    <code>true</code> if <code>s</em> matches the regular expression;
    *    <code>false</code> if it does not.
    *
    * @throws IllegalArgumentException
    *    if <code>regex == null</code> or if it has an invalid syntax. 
    */
   private static final boolean matches(String s, String regex)
   throws IllegalArgumentException {

      // Check preconditions
      if (regex == null) {
         throw new IllegalArgumentException("regex == null");
      }

      // Compile the regular expression pattern
      Pattern pattern;
      try {
         pattern = Pattern.compile(regex);
      } catch (PatternSyntaxException cause) {
         throw new IllegalArgumentException("Invalid regular expression \"" + regex + "\".", cause);
      }

      // Short-circuit if the string is null
      if (s == null) {
         return false;
      }

      // Find a match
      return pattern.matcher(s).find();
   }

   /**
    * Checks if the specified string is either null or empty (after trimming
    * the whitespace off).
    *
    * @param s
    *    the string to check.
    *
    * @return
    *    <code>true</code> if <code>s == null || s.trim().length() &lt; 1</code>;
    *    <code>false</code> otherwise.
    */
   private static final boolean isEmpty(String s) {
      return s == null || s.trim().length() < 1;
   }

   /**
    * Checks if the specified abstract path name refers to an existing
    * directory.
    *
    * @param description
    *    the description of the directory, cannot be <code>null</code>.
    *
    * @param path
    *    the abstract path name as a {@link File} object.
    *
    * @param mustBeReadable
    *    <code>true</code> if the directory must be readable.
    *
    * @param mustBeWritable
    *    <code>true</code> if the directory must be writable.
    *
    * @throws IllegalArgumentException
    *    if <code>location == null
    *          || {@linkplain TextUtils}.{@linkplain TextUtils#isEmpty(String) isEmpty}(description)</code>.
    *
    * @throws BuildException
    *    if <code>  path == null
    *          || ! path.exists()
    *          || ! path.isDirectory()
    *          || (mustBeReadable &amp;&amp; !path.canRead())
    *          || (mustBeWritable &amp;&amp; !path.canWrite())</code>.
    */
   private static final void checkDir(String  description,
                                      File    path,
                                      boolean mustBeReadable,
                                      boolean mustBeWritable)
   throws IllegalArgumentException, BuildException {

      // Check preconditions
      if (isEmpty(description)) {
         throw new IllegalArgumentException("description is empty (" + quote(description) + ')');
      }

      // Make sure the path refers to an existing directory
      if (path == null) {
         throw new BuildException(description + " is not set.");
      } else if (! path.exists()) {
         throw new BuildException(description + " (\"" + path + "\") does not exist.");
      } else if (! path.isDirectory()) {
         throw new BuildException(description + " (\"" + path + "\") is not a directory.");

      // Make sure the directory is readable, if that is required
      } else if (mustBeReadable && (! path.canRead())) {
         throw new BuildException(description + " (\"" + path + "\") is not readable.");

      // Make sure the directory is writable, if that is required
      } else if (mustBeWritable && (! path.canWrite())) {
         throw new BuildException(description + " (\"" + path + "\") is not writable.");
      }
   }


   //-------------------------------------------------------------------------
   // Constructors
   //-------------------------------------------------------------------------

   /**
    * Constructs a new <code>OptiPNGTask</code> object.
    */
   public OptiPNGTask() {
      // empty
   }


   //-------------------------------------------------------------------------
   // Fields
   //-------------------------------------------------------------------------

   /**
    * The directory to read the image files from.
    * See {@link #setDir(File)}.
    */
   private File _sourceDir;

   /**
    * The directory to write the PNG image files to.
    * See {@link #setToDir(File)}.
    */
   private File _destDir;

   /**
    * Flag that indicates if each existing file should always be overwritten,
    * even if it is newer than the source file. Default is <code>false</code>.
    */
   private boolean _overwrite;

   /**
    * The command to execute. If unset, then this task will attempt to find a
    * proper executable by itself.
    */
   private String _command;

   /**
    * The time-out to apply, in milliseconds, or 0 (or lower) in case no
    * time-out should be applied.
    */
   private long _timeOut;

   /**
    * Character string that indicates whether the files should be processed
    * with OptiPNG at all. There are 3 options:
    * <dl>
    * <dt><code>"yes"</code> or <code>"true"</code>
    * <dd>The files <em>must</em> be processed by OptiPNG.
    *     If at least one file could not be processed, for example because
    *     OptiPNG is unavailable or because there are unrecoverable errors,
    *     then this task fails.
    *
    * <dt><code>"no"</code> or <code>"false"</code>
    * <dd>The files must <em>not</em> be processed by OptiPNG but must instead
    *     just be copied as-is, unchanged.
    *
    * <dt><code>"try"</code>
    * <dd>Attempt to process the file, but if that fails, then just copy the
    *     original file instead.
    * </dl>
    */
   private String _process;

   
   //-------------------------------------------------------------------------
   // Methods
   //-------------------------------------------------------------------------

   /**
    * Sets the path to the source directory. This parameter is required.
    *
    * @param dir
    *    the location of the source directory, or <code>null</code>.
    */
   public void setDir(File dir) {
      log("Setting \"dir\" to: " + quote(dir) + '.', MSG_VERBOSE);
      _sourceDir = dir;
   }

   /**
    * Sets the path to the destination directory. The default is the same
    * directory.
    *
    * @param dir
    *    the location of the destination directory, or <code>null</code>.
    */
   public void setToDir(File dir) {
      log("Setting \"toDir\" to: " + quote(dir) + '.', MSG_VERBOSE);
      _destDir = dir;
   }

   /**
    * Sets the <em>overwrite</em> flag.
    *
    * @param flag
    *    the value for the flag.
    */
   public void setOverwrite(boolean flag) {
      _overwrite = flag;
   }

   /**
    * Sets the command to execute, optionally. By default this task will find
    * a proper command on the current path.
    *
    * @param command
    *    the command to use, e.g. <code>"/usr/local/bin/optipng"</code>,
    *    can be <code>null</code> (in which case the task will find the command).
    */
   public void setCommand(String command) {
      log("Setting \"command\" to: " + quote(command) + '.', MSG_VERBOSE);
      _command = command;
   }

   /**
    * Configures the time-out for executing a single OptiPNG command. The
    * default is 60 seconds. Setting this to 0 or lower disables the time-out
    * completely.
    *
    * @param timeOut
    *    the time-out to use in milliseconds, or 0 (or lower) if no time-out
    *    should be applied.
    */
   public void setTimeOut(long timeOut) {
      log("Setting \"timeOut\" to: " + timeOut + " ms.", MSG_VERBOSE);
      _timeOut = timeOut;
   }

   /**
    * Sets whether the files should be processed with OptiPNG at all.
    * There are 3 options:
    * <dl>
    * <dt><code>"yes"</code> or <code>"true"</code>
    * <dd>The files <em>must</em> be processed by OptiPNG.
    *     If OptiPNG is unavailable, then this task fails.
    *
    * <dt><code>"no"</code> or <code>"false"</code>
    * <dd>The files must <em>not</em> be processed by OptiPNG but must instead
    *     just be copied as-is, unchanged.
    *
    * <dt><code>"try"</code>
    * <dd>If OptiPNG is available process the file(s), but if OptiPNG is
    *     unavailable, then just copy the files instead (in which case a
    *     warning will be output).
    * </dl>
    *
    * @param s
    *    the value, should be one of the allowed values (otherwise the task
    *    will fail during execution).
    */
   public void setProcess(String s) {
      log("Setting \"process\" to: " + quote(s) + '.', MSG_VERBOSE);
      _process = s;
   }

   @Override
   public void execute() throws BuildException {

      // Source directory defaults to current directory
      if (_sourceDir == null) {
         _sourceDir = getProject().getBaseDir();
      }

      // Destination directory defaults to source directory
      if (_destDir == null) {
         _destDir = _sourceDir;
      }

      // Check the directories
      checkDir("Source directory",      _sourceDir,  true, false);
      checkDir("Destination directory",   _destDir, false,  true);

      // Interpret the "process" option
      ProcessOption processOption;
      String p = (_process == null) ? null : _process.toLowerCase().trim();
      if (p == null || "true".equals(p) || "yes".equals(p)) {
         processOption = ProcessOption.MUST;
      } else if ("false".equals(_process) || "no".equals(_process)) {
         processOption = ProcessOption.MUST_NOT;
      } else if ("try".equals(_process)) {
         processOption = ProcessOption.SHOULD;
      } else {
         throw new BuildException("Invalid value for \"process\" option: " + quote(_process) + '.');
      }

      // Determine what command to execute
      String command = (_command == null || _command.length() < 1)
                     ? DEFAULT_COMMAND
                     : _command;

      // Test that the command is available
      boolean commandAvailable = testCommand(command, processOption);

      // Determine if transformation should be attempted at all
      // (alternative is just copying)
      boolean transform = processOption != ProcessOption.MUST_NOT && commandAvailable;

      // Consider each individual file for processing/copying
      log("Transforming from " + _sourceDir.getPath() + " to " + _destDir.getPath() + '.', MSG_VERBOSE);
      long start = System.currentTimeMillis();
      int failedCount = 0, optimizeCount = 0, copyCount = 0, skippedCount = 0;
      for (String inFileName : getDirectoryScanner(_sourceDir).getIncludedFiles()) {

         // Make sure the input file exists
         File inFile = new File(_sourceDir, inFileName);
         if (! inFile.exists()) {
            continue;
         }

         // Determine if the file type is supported
         if (! matches(inFileName.toLowerCase(), "\\.png$")) {
            log("Skipping " + quote(inFileName) + " because the file does not end in \".png\" (case-insensitive).", MSG_VERBOSE);
            skippedCount++;
            continue;
         }

         // Some preparations related to the input file and output file
         long     thisStart = System.currentTimeMillis();
         String outFileName = inFileName.replaceFirst("\\.[a-zA-Z]+$", ".png");
         File       outFile = new File(_destDir, outFileName);
         String outFilePath = outFile.getPath();
         String  inFilePath = inFile.getPath();

         // Skip this file is the output file exists and is newer
         if (!_overwrite && outFile.exists() && (outFile.lastModified() > inFile.lastModified())) {
            log("Skipping " + quote(inFileName) + " because output file is newer.", MSG_VERBOSE); 
            skippedCount++;
            continue;

         // Skip each empty file
         } else if (inFile.length() < 1L) {
            log("Skipping " + quote(inFileName) + " because the file is completely empty.", MSG_VERBOSE); 
            skippedCount++;
            continue;
         }

         // File transformation (optimization) should be attempted
         boolean copy = !transform;
         if (transform) {

            // Prepare for the command execution
            Buffer            buffer = new Buffer();
            ExecuteWatchdog watchdog = (_timeOut > 0L) ? new ExecuteWatchdog(_timeOut) : null;
            Execute          execute = new Execute(buffer, watchdog);
            String[]         cmdline = new String[] { command, "-fix", "-force", "-out", outFilePath, "--", inFilePath };

            execute.setAntRun(getProject());
            execute.setCommandline(cmdline);

            // Execute the command
            boolean failure;
            try {
               execute.execute();
               failure = execute.isFailure();
            } catch (IOException cause) {
               failure = true;
            }

            // Output to stderr indicates a failure
            String errorOutput = buffer.getErrString();
            failure            = failure ? true : ! isEmpty(errorOutput);

            // A non-existent or empty file also indicate failure
            if (! failure) {
               if (! outFile.exists()) {
                  failure     = true;
                  errorOutput = "Output file not created.";
               } else if (outFile.length() < 1L) {
                  failure     = true;
                  errorOutput = "Generated output file is empty.";
               }
            }

            // Log the result for this individual file
            long thisDuration = System.currentTimeMillis() - thisStart;
            if (failure) {
               String logMessage = "Failed to optimize " + quote(inFilePath);
               if (isEmpty(errorOutput)) {
                  logMessage += '.';
               } else {
                  logMessage += ": " + errorOutput;
               }
               log(logMessage, MSG_ERR);
               failedCount++;

               // Failed, but then instead copy the input file unchanged
               if (processOption != ProcessOption.MUST) {
                  copy = true;
               }
            } else {
               log("Optimized " + quote(inFileName) + " in " + thisDuration + " ms.", MSG_VERBOSE);
               optimizeCount++;
            }
         }

         // Copy the file?
         if (copy) {
            try {
               FileUtils.getFileUtils().copyFile(inFile, outFile);
               long thisDuration = System.currentTimeMillis() - thisStart;
               log("Copied " + quote(inFileName) + " in " + thisDuration + " ms.", MSG_VERBOSE);
               copyCount++;
            } catch (Throwable cause) {
               String logMessage = "Failed to copy " + quote(inFilePath) + " to " + quote(outFilePath) + '.';
               log(logMessage, MSG_ERR);
               failedCount++;
            }
         }
      }

      // Log the total result
      long duration = System.currentTimeMillis() - start;
      if (failedCount > 0) {
         throw new BuildException("" + failedCount + " file(s) failed to be optimized and/or copied; " + optimizeCount + " file(s) optimized; " + copyCount + " file(s) copied; " + skippedCount + " file(s) skipped. Total duration is " + duration + " ms.");
      } else {
         log("" + optimizeCount + " file(s) optimized and " + copyCount + " file(s) copied in " + duration + " ms; " + skippedCount + " file(s) skipped.");
      }
   }

   private boolean testCommand(String command, ProcessOption processOption)
   throws IllegalArgumentException, BuildException {

      // Check preconditions
      if (command == null) {
         throw new IllegalArgumentException("command == null");
      } else if (processOption == null) {
         throw new IllegalArgumentException("processOption == null");
      }

      // Short-circuit if no command should be executed
      if (processOption == ProcessOption.MUST_NOT) {
         return false;
      }

      // Create a watch dog, if a time-out is configured
      ExecuteWatchdog watchdog = (_timeOut > 0L) ? new ExecuteWatchdog(_timeOut) : null;

      // Check that the command is executable
      Buffer    buffer = new Buffer();
      Execute  execute = new Execute(buffer, watchdog);
      String[] cmdline = new String[] { command, "-version" };
      execute.setAntRun(getProject());
      execute.setCommandline(cmdline);
      Throwable caught;
      try {
         execute.execute();
         caught = null;
      } catch (Throwable e) {
         caught = e;
      }

      // Executing the command triggered an exception
      boolean commandAvailable;
      if (caught != null) {
         String message = "Unable to execute OptiPNG command " + quote(command) + '.';
         if (processOption == ProcessOption.MUST) {
            throw new BuildException(message, caught);
         } else {
            log(message, MSG_ERR);
            commandAvailable = false;
         }

      // Executing the command resulted in a non-zero code, indicating failure
      } else if (execute.getExitValue() != 0) {
         String message = "Unable to execute OptiPNG command " + quote(command) + ". Running '" + command + " -v' resulted in exit code " + execute.getExitValue() + '.';
         if (processOption == ProcessOption.MUST) {
            throw new BuildException(message);
         } else {
            log(message, MSG_ERR);
            commandAvailable = false;
         }

      // Command was executed successfully
      } else {
         Pattern pattern = Pattern.compile("^[^0-9]*([0-9]+(\\.[0-9]+)*)");
         Matcher matcher = pattern.matcher(buffer.getOutString());
         String  version = matcher.find() ? quote(matcher.group(1)) : "unknown";
         log("Using command " + quote(command) + ", version is " + version + '.', MSG_VERBOSE);
         commandAvailable = true;
      }

      return commandAvailable;
   }


   //-------------------------------------------------------------------------
   // Inner classes
   //-------------------------------------------------------------------------

   /**
    * Enumeration type for the different process options.
    *
    * @author <a href="mailto:ernst@pensioenpage.com">Ernst de Haan</a>
    */
   private enum ProcessOption {

      /**
       * Force processing with OptiPNG. If the OptiPNG command is not
       * available, then fail.
       */
      MUST,
         
      /**
       * Skip OptiPNG processing completely. Just copy the files.
       */
      MUST_NOT,

      /**
       * Try OptiPNG processing. If the processing fails, then copy the
       * original file.
       */
      SHOULD;
   }
}
